x-pack/filebeat/input/http_endpoint/handler.go

```diff
@@ -82,34 +93,94 @@ func sendErrorResponse(w http.ResponseWriter, status int, err error) {
 }
 
-func httpReadJSON(body io.Reader) (objs []common.MapStr, status int, err error) {
+func httpReadJSON(body io.Reader) (objs []common.MapStr, rawMessages []json.RawMessage, status int, err error) {
 	if body == http.NoBody {
-		return nil, http.StatusNotAcceptable, errBodyEmpty
+		return nil, nil, http.StatusNotAcceptable, errBodyEmpty
 	}
+	obj, rawMessage, err := decodeJSON(body)
+	if err != nil {
+		return nil, nil, http.StatusBadRequest, err
+	}
+	return obj, rawMessage, http.StatusOK, err
+
+}
 
+func decodeJSON(body io.Reader) (objs []common.MapStr, rawMessages []json.RawMessage, err error) {
 	decoder := json.NewDecoder(body)
-	for idx := 0; ; idx++ {
-		var obj interface{}
-		if err := decoder.Decode(&obj); err != nil {
+	for decoder.More() {
+		var raw json.RawMessage
+		if err := decoder.Decode(&raw); err != nil {
 			if err == io.EOF {
 				break
 			}
-			return nil, http.StatusBadRequest, errors.Wrapf(err, "malformed JSON object at stream position %d", idx)
+			return nil, nil, errors.Wrapf(err, "malformed JSON object at stream position %d", decoder.InputOffset())
+		}
+
+		var obj interface{}
+		if err := newJSONDecoder(bytes.NewReader(raw)).Decode(&obj); err != nil {
+			return nil, nil, errors.Wrapf(err, "malformed JSON object at stream position %d", decoder.InputOffset())
 		}
 		switch v := obj.(type) {
 		case map[string]interface{}:
 			objs = append(objs, v)
+			rawMessages = append(rawMessages, raw)
 		case []interface{}:
-			for listIdx, listObj := range v {
-				asMap, ok := listObj.(map[string]interface{})
-				if !ok {
-					return nil, http.StatusBadRequest, fmt.Errorf("%v at stream %d index %d", errUnsupportedType, idx, listIdx)
-				}
-				objs = append(objs, asMap)
+			nobjs, nrawMessages, err := decodeJSONArray(bytes.NewReader(raw))
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "recursive error %d", decoder.InputOffset())
 			}
+			objs = append(objs, nobjs...)
+			rawMessages = append(rawMessages, nrawMessages...)
 		default:
-			return nil, http.StatusBadRequest, errUnsupportedType
+			return nil, nil, errUnsupportedType
+		}
+	}
+	return objs, rawMessages, nil
+}
+
+func decodeJSONArray(raw *bytes.Reader) (objs []common.MapStr, rawMessages []json.RawMessage, err error) {
+	dec := newJSONDecoder(raw)
+	token, err := dec.Token()
+	if token != json.Delim('[') || err != nil {
+		return nil, nil, errors.Wrapf(err, "malformed JSON array, not starting with delimiter [ at position: %d", dec.InputOffset())
+	}
+
+	for dec.More() {
+		var raw json.RawMessage
+		if err := dec.Decode(&raw); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, nil, errors.Wrapf(err, "malformed JSON object at stream position %d", dec.InputOffset())
+		}
+
+		var obj interface{}
+		if err := newJSONDecoder(bytes.NewReader(raw)).Decode(&obj); err != nil {
+			return nil, nil, errors.Wrapf(err, "malformed JSON object at stream position %d", dec.InputOffset())
+		}
+
+		rawMessages = append(rawMessages, raw)
```

Comment:
To be sure the size of the returned slices are always the same I think this should be moved down to where `objs` is appended to.